client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -292,20 +298,48 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
   }
 
   // transform all [partition, records] to [partition, ShuffleBlockInfo] and clear cache
-  public synchronized List<ShuffleBlockInfo> clear() {
+  public synchronized List<ShuffleBlockInfo> clear(double bufferSpillRatio) {
     List<ShuffleBlockInfo> result = Lists.newArrayList();
     long dataSize = 0;
     long memoryUsed = 0;
-    Iterator<Entry<Integer, WriterBuffer>> iterator = buffers.entrySet().iterator();
-    while (iterator.hasNext()) {
-      Entry<Integer, WriterBuffer> entry = iterator.next();
-      WriterBuffer wb = entry.getValue();
+    bufferSpillRatio = Math.max(0.1, Math.min(1.0, bufferSpillRatio));
+    List<Integer> partitionList =
+        new ArrayList<Integer>() {
+          {
+            addAll(buffers.keySet());
+          }
+        };
+    if (bufferSpillRatio < 1.0) {
+      Collections.sort(
```

Comment:
This can be simplified as:
```
partitionList.sort(Comparator.comparingInt(o -> buffers.get(o) == null ? 0 : buffers.get(o).getMemoryUsed()).reversed());
```